File: convert/singbox/dns.go
package singbox

import (
	"encoding/json"
	"singboxconfig/entity"

	"github.com/sirupsen/logrus"
)

// GetDefaultDNS 返回与历史硬编码一致的默认 DNS 配置。
// 当存储里没有 dns_config，或保存值非法时，生成链路会回退到这里。
func GetDefaultDNS() entity.SingDNS {
	dnsServers := []entity.SingDNSServer{
		{
			Tag:             "dns_proxy",
			Address:         "https://1.1.1.1/dns-query",
			AddressResolver: "dns_resolver",
			Strategy:        "ipv4_only",
			Detour:          "general",
		},
		{
			Tag:             "dns_direct",
			Address:         "h3://dns.alidns.com/dns-query",
			AddressResolver: "dns_resolver",
			Strategy:        "ipv4_only",
			Detour:          "direct",
		},
		{
			Tag:     "dns_block",
			Address: "rcode://refused",
		},
		{
			Tag:      "dns_resolver",
			Address:  "223.5.5.5",
			Strategy: "ipv4_only",
			Detour:   "direct",
		},
	}

	dnsRules := []entity.SingDNSRule{
		{
			Outbound: "any",
			Server:   "dns_resolver",
		},
		{
			ClashMode: "direct",
			Server:    "dns_direct",
		},
		{
			ClashMode: "global",
			Server:    "dns_proxy",
		},
		{
			RuleSet: "cnip",
			Server:  "dns_direct",
		},
		{
			RuleSet: "cnsite",
			Server:  "dns_direct",
		},
	}

	return entity.SingDNS{
		Servers: dnsServers,
		Rules:   dnsRules,
		Final:   "dns_proxy",
	}
}

// ResolveDNS 按存储值解析 DNS 配置。
// 只要 JSON 为空或不合法，就统一回退到默认 DNS，避免生成接口因配置损坏直接失败。
func ResolveDNS(configJSON string) entity.SingDNS {
	if configJSON == "" {
		return GetDefaultDNS()
	}

	var dns entity.SingDNS
	if err := json.Unmarshal([]byte(configJSON), &dns); err != nil {
		logrus.WithError(err).Warn("ResolveDNS: invalid dns_config, fallback to default")
		return GetDefaultDNS()
	}

	return dns
}

File: web/src/utils/dns.ts
import type { SingDNS } from '../types';

export const DNS_SETTING_KEY = 'dns_config';

// 与当前后端 convert/singbox/dns.go 的默认内容保持一致，用于“恢复默认”按钮。
export const DEFAULT_DNS_CONFIG: SingDNS = {
  servers: [
    {
      tag: 'dns_proxy',
      address: 'https://1.1.1.1/dns-query',
      address_resolver: 'dns_resolver',
      strategy: 'ipv4_only',
      detour: 'general',
    },
    {
      tag: 'dns_direct',
      address: 'h3://dns.alidns.com/dns-query',
      address_resolver: 'dns_resolver',
      strategy: 'ipv4_only',
      detour: 'direct',
    },
    {
      tag: 'dns_block',
      address: 'rcode://refused',
    },
    {
      tag: 'dns_resolver',
      address: '223.5.5.5',
      strategy: 'ipv4_only',
      detour: 'direct',
    },
  ],
  rules: [
    {
      outbound: 'any',
      server: 'dns_resolver',
    },
    {
      clash_mode: 'direct',
      server: 'dns_direct',
    },
    {
      clash_mode: 'global',
      server: 'dns_proxy',
    },
    {
      rule_set: 'cnip',
      server: 'dns_direct',
    },
    {
      rule_set: 'cnsite',
      server: 'dns_direct',
    },
  ],
  final: 'dns_proxy',
};

export function getDefaultDnsEditorText(): string {
  return JSON.stringify(DEFAULT_DNS_CONFIG, null, 2);
}

export function getDnsEditorText(settingValue?: string | null): string {
  if (!settingValue) {
    return getDefaultDnsEditorText();
  }

  try {
    return JSON.stringify(JSON.parse(settingValue), null, 2);
  } catch {
    return settingValue;
  }
}
